File: matlab/gpOptimise.m
function model = gpOptimise(model, display, iters,gradcheck);

% GPOPTIMISE Optimise the inducing variable based kernel.
% FORMAT
% DESC optimises the Gaussian process  model for
% a given number of iterations.
% ARG model : the model to be optimised.
% ARG display : whether or not to display while optimisation
% proceeds, set to 2 for the most verbose and 0 for the least
% verbose.
% ARG iters : number of iterations for the optimisation.
% RETURN model : the optimised model.
%
% SEEALSO : scg, conjgrad, gpCreate, gpGradient, gpObjective
%
% COPYRIGHT : Neil D. Lawrence, 2005, 2006
%
% MODIFICATIONS : Carl Henrik Ek, 2008

% GP

if(nargin<4)
  gradcheck = false;
  if nargin < 3
    iters = 2000;
    if nargin < 2
      display = 1;
    end
  end
end


params = gpExtractParam(model);

options = optOptions;
if display
  options(1) = 1;
  if length(params) <= 100 && gradcheck
    options(9) = 1;
  end
end
options(14) = iters;

if isfield(model, 'optimiser')
  optim = str2func(model.optimiser);
else
  optim = str2func('conjgrad');
end

if strcmp(func2str(optim), 'optimiMinimize')
  % Carl Rasmussen's minimize function 
  params = optim('gpObjectiveGradient', params, options, model);
else
  % NETLAB style optimization.
  params = optim('gpObjective', params,  options, ...
                 'gpGradient', model);
end

model = gpExpandParam(model, params);

File: R/gpOptimise.R
gpOptimise <- function(model, display=TRUE, iters=2000, gradcheck=FALSE) {

#   if (nargs() < 4) {
#     gradcheck = FALSE
#     if (nargs() < 3) {
#       iters = 2000
#       if (nargs() < 2)
# 	display = TRUE
#     }
#   }

  params = gpExtractParam(model)
  ## options = list(maxit=3000, ln=c(0,2), xtol=1e-4, fnTol=1e-4, optimiser="SCG",
  ##	gradcheck=FALSE, display=TRUE)
  options = optimiDefaultOptions() #optOptions
  if (display) {
    options$display = TRUE
    if ((length(params) <= 100) && gradcheck)
      options$gradcheck = TRUE
  }
  options$maxit = iters

  if ("optimiser" %in% names(model))
    optim = get(paste(model$optimiser, "optim", sep=""), mode="function")
  else
    optim = get("CGoptim", mode="function")

  fn = get('gpObjective', mode="function")
  grad = get('gpGradient', mode="function")

#   strcmp(func2str(optim), 'optimiMinimize')
#     ## Carl Rasmussen's minimize function 
#     params = optim('gpObjectiveGradient', params, options, model);
#   else
  ## R version of NETLAB function
  ## only necessary parts from gpGradient implemented!
  params = optim(params, fn, grad, options, model) ## !!!

  model = gpExpandParam(model, params$xmin)

  return (model)
}